core/accounting/src/fiscal_year/entity.rs

```diff
@@ -0,0 +1,223 @@
+use chrono::{DateTime, Datelike, Months, NaiveDate, Utc};
+use derive_builder::Builder;
+use es_entity::*;
+#[cfg(feature = "json-schema")]
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use super::error::FiscalYearError;
+use crate::primitives::{ChartId, FiscalYearId};
+
+#[derive(EsEvent, Debug, Clone, Serialize, Deserialize)]
+#[cfg_attr(feature = "json-schema", derive(JsonSchema))]
+#[serde(tag = "type", rename_all = "snake_case")]
+#[es_event(id = "FiscalYearId")]
+pub enum FiscalYearEvent {
+    Initialized {
+        id: FiscalYearId,
+        chart_id: ChartId,
+        first_period_opened_as_of: chrono::NaiveDate,
+        first_period_opened_at: chrono::DateTime<Utc>,
+    },
+    MonthClosed {
+        closed_as_of: NaiveDate,
+        closed_at: DateTime<Utc>,
+    },
+}
+
+#[derive(EsEntity, Builder, Clone)]
+#[builder(pattern = "owned", build_fn(error = "EsEntityError"))]
+pub struct FiscalYear {
+    pub id: FiscalYearId,
+    pub chart_id: ChartId,
+    pub first_period_opened_as_of: NaiveDate,
+    pub first_period_opened_at: DateTime<Utc>,
+    #[builder(default)]
+    pub last_month_closed_at: Option<DateTime<Utc>>,
+
+    events: EntityEvents<FiscalYearEvent>,
+}
+
+impl FiscalYear {
+    pub fn close_last_month(
+        &mut self,
+        now: DateTime<Utc>,
+    ) -> Result<Idempotent<NaiveDate>, FiscalYearError> {
+        let last_recorded_date = self.events.iter_all().rev().find_map(|event| match event {
+            FiscalYearEvent::MonthClosed { closed_as_of, .. } => Some(*closed_as_of),
+            _ => None,
+        });
+        let new_monthly_closing_date = match last_recorded_date {
+            Some(last_effective) => {
+                let last_day_of_previous_month = now
+                    .date_naive()
+                    .with_day(1)
+                    .and_then(|d| d.pred_opt())
+                    .expect("Failed to compute last day of previous month");
+                if last_effective == last_day_of_previous_month {
+                    return Ok(Idempotent::Ignored);
+                }
+
+                last_effective
+                    .checked_add_months(Months::new(2))
+                    .and_then(|d| d.with_day(1))
+                    .and_then(|d| d.pred_opt())
+                    .expect("Failed to compute new monthly closing date")
+            }
+            None => self
+                .events
+                .iter_all()
+                .find_map(|event| match event {
+                    FiscalYearEvent::Initialized {
+                        first_period_opened_as_of,
+                        ..
+                    } => Some(*first_period_opened_as_of),
+                    _ => None,
+                })
+                .ok_or(FiscalYearError::InitFiscalYearNotFound)?
```

Comment:
We probably won't ever hit this `ok_or`, can maybe use an `expect` instead
```suggestion
                .expect("Entity was not initialized")
```